refactor(compiler-cli): implement `ɵɵngDeclareNgModule` and `ɵɵngDeclareInjector`

petebacondarwin opened the pull request:
This commit changes the partial compilation so that it outputs declaration
calls rather than definition calls for NgModules and Injectors.

The JIT compiler and the linker are updated to be able to handle these
new declarations.

---


This sits on top of #41022 so only commits after **refactor(compiler-cli): implement `ɵɵngDeclareNgModule` and `ɵɵngDeclareInjector`** need to be considered.

---

Note that this creates a new LinkerOption called `linkerJitMode`, which I think needs to be made available on `tsconfig.json` `angularCompilerOptions`. *Open for discussion*.

petebacondarwin committed:
refactor(compiler-cli): implement `ɵɵngDeclareNgModule` and `ɵɵngDeclareInjector`

This commit changes the partial compilation so that it outputs declaration
calls rather than definition calls for NgModules and Injectors.

The JIT compiler and the linker are updated to be able to handle these
new declarations.

petebacondarwin commented:
This PR is blocked on top of #41022

alxhub commented on packages/compiler-cli/linker/src/file_linker/linker_options.ts:
```diff
@@ -33,6 +33,8 @@ export interface LinkerOptions {
    */
   sourceMapping: boolean;
+
+  linkerJitMode: boolean;
```
Please sir, may I have some docs?

petebacondarwin replied:
Ah yes - I didn't document this because I was wondering about how we were going to expose it and then I forgot about it. Let's have a chat in our Monday sync. I think this could either be explicit in tsconfig.json: `angularCompilerOptions` or provided by the CLI based on the angular.json `aot: false` mode.

petebacondarwin replied:
Added some docs.

alxhub commented on packages/compiler-cli/linker/src/file_linker/partial_linkers/partial_injector_linker_1.ts:
```diff
@@ -0,0 +1,51 @@
+ * A `PartialLinker` that is designed to process `ɵɵngDeclareInjector()` call expressions.
+ */
+export class PartialInjectorLinkerVersion1<TExpression> implements PartialLinker<TExpression> {
+  constructor() {}
```
Nit: unnecessary?

petebacondarwin replied:
Indeed - at some point it had a parameter :-)

alxhub commented on packages/compiler-cli/linker/src/file_linker/partial_linkers/partial_ng_module_linker_1.ts:
```diff
@@ -0,0 +1,65 @@
+ * A `PartialLinker` that is designed to process `ɵɵngDeclareNgModule()` call expressions.
+ */
+export class PartialNgModuleLinkerVersion1<TExpression> implements PartialLinker<TExpression> {
+  constructor(private emitInline: boolean = false) {}
```
This feels like the kind of parameter that should be:

1) required, not optional with a default value.
2) an enum instead of a boolean, with documented variants that describe the effect on code generation

petebacondarwin replied:
Indeed! I think I made it optional while hacking to save having to update all the code.

petebacondarwin replied:
I am not convinced that it should be an enum. The `emitInline` is just passed through to the compiler, which already has this option (with the same name) as a boolean.

alxhub commented on packages/compiler-cli/src/ngtsc/annotations/src/ng_module.ts:
```diff
@@ -6,8 +6,10 @@
-import {compileInjector, compileNgModule, CUSTOM_ELEMENTS_SCHEMA, Expression, ExternalExpr, Identifiers as R3, InvokeFunctionExpr, LiteralArrayExpr, LiteralExpr, NO_ERRORS_SCHEMA, R3FactoryTarget, R3Identifiers, R3InjectorMetadata, R3NgModuleMetadata, R3Reference, SchemaMetadata, Statement, STRING_TYPE, WrappedNodeExpr} from '@angular/compiler';
-import {R3DependencyMetadata} from '@angular/compiler/src/compiler';
+import {compileDeclareInjectorFromMetadata, compileDeclareNgModuleFromMetadata, compileFactoryFunction, compileInjector, compileNgModule, CUSTOM_ELEMENTS_SCHEMA, Expression, ExternalExpr, Identifiers as R3, InvokeFunctionExpr, LiteralArrayExpr, LiteralExpr, NO_ERRORS_SCHEMA, R3FactoryMetadata, R3FactoryTarget, R3Identifiers, R3InjectorMetadata, R3NgModuleMetadata, R3Reference, SchemaMetadata, Sta ...
+import {R3FactoryFn} from '@angular/compiler/src/render3/r3_factory';
```
These symbols should probably be exported from the top level of `@angular/compiler`.

On a semi-related note, we should probably sit down and actually _define_ (hehe) an API surface for `@angular/compiler`, besides just exporting everything as an ES grab bag.

petebacondarwin replied:
Yes and Yes.

petebacondarwin replied:
> On a semi-related note, we should probably sit down and actually define (hehe) an API surface for @angular/compiler, besides just exporting everything as an ES grab bag.

We can add it to the backlog of semi-public APIs that we want to nail down.

AndrewKushnir reviewed (approved):
Reviewed-for: public-api
Reviewed-for: size-tracking

The only file affected by the change from the `public-api` standpoint: `goldens/public-api/core/core.d.ts`

(similar to #41022)

jessicajaniuk commented on packages/compiler/src/render3/r3_module_compiler.ts:
```diff
@@ -8,9 +8,8 @@
-import {compileFactoryFunction, R3DependencyMetadata, R3FactoryTarget} from './r3_factory';
 import {Identifiers as R3} from './r3_identifiers';
-import {jitOnlyGuardedExpression, mapToMapExpression, R3Reference} from './util';
+import {jitOnlyGuardedExpression, mapToMapExpression, R3Reference, refsToArray} from './util';
```
I see this added import, but it appears to not be used in any of the added code. Am I missing something?

petebacondarwin replied:
It is actually used around line 118. Originally the function was defined at the bottom of this file. But now it is imported.

jessicajaniuk reviewed (approved):
Just a couple of nits / questions, but otherwise looks good. LGTM!

reviewed-for: public-api

petebacondarwin commented on packages/compiler-cli/src/ngtsc/annotations/src/ng_module.ts:
```diff
@@ -227,7 +226,7 @@ export class NgModuleDecoratorHandler implements
     const adjacentType = new WrappedNodeExpr(this.reflector.getAdjacentNameOfClass(node));
 
-    const ngModuleDef: R3NgModuleMetadata = {
+    const mod: R3NgModuleMetadata = {
```
I'll update the variable names.

JoostK commented on packages/compiler-cli/linker/src/file_linker/partial_linkers/partial_ng_module_linker_1.ts:
```diff
@@ -0,0 +1,70 @@
+export class PartialNgModuleLinkerVersion1<TExpression> implements PartialLinker<TExpression> {
+  constructor(
+      /**
+         If true then emit the additional declarations, imports, exports, etc in the NgModule
```
Your editor doesn't appear to like asterisks, does it...

petebacondarwin replied:
What?!?!?

JoostK commented on packages/compiler-cli/linker/src/file_linker/partial_linkers/util.ts:
```diff
@@ -0,0 +1,13 @@
+import {R3Reference} from '@angular/compiler';
+import * as o from '@angular/compiler/src/output/output_ast';
+
+export function wrapReference<TExpression>(wrapped: o.WrappedNodeExpr<TExpression>): R3Reference {
```
Did you mean to use this in `partial_ng_module_linker` instead of its `toR3Reference`?

JoostK commented on packages/compiler-cli/linker/src/file_linker/partial_linkers/partial_ng_module_linker_1.ts:
```diff
@@ -0,0 +1,70 @@
+    exports: getWrappedArray(metaObj, 'exports'),
+    emitInline,
+    containsForwardDecls:
+        metaObj.has('containsForwardDecls') && metaObj.getBoolean('containsForwardDecls'),
```
I am surprised this is part of the partial declaration; couldn't we figure this out from e.g. a function wrapper? Additionally, I am wondering if this could result in double wrapping, given that the list that contains forward decls has already been wrapped? But that can't be the case currently as `getWrappedArray` always reads an array and never a function... I'm probably missing something here.

petebacondarwin replied:
As you note later. The idea is that the declaration interface basically exposes the `R3NgModuleMeta` information, which we then pass through to a compile function. This keeps the compilation neat, since both full compile and linker use the same input metadata.

We could somehow split this up so that the partial compilation partly processes this metadata so the linker has a little less to do, but this will make the code more complex for little real benefit in this case. It also allows us to hide exactly how we handle forward references in these cases, since it delays the implementation of that until linking time rather than baking it into what is published in the library.

JoostK commented on packages/compiler-cli/src/ngtsc/annotations/src/ng_module.ts:
```diff
@@ -450,17 +459,58 @@ export class NgModuleDecoratorHandler implements
+      node: ClassDeclaration, {inj, fac, mod, metadataStmt}: Readonly<NgModuleAnalysis>,
+      {injectorImports}: Readonly<NgModuleResolution>): CompileResult[] {
+    const factoryFn = compileFactoryFunction(fac);
+    if (metadataStmt !== null) {
```
Should this use `appendMetadataStatement`?

JoostK commented on packages/compiler-cli/src/ngtsc/annotations/src/ng_module.ts:
```diff
@@ -450,17 +459,58 @@ export class NgModuleDecoratorHandler implements
+    this.appendMetadataStatement(ngModuleDef.additionalStatements, metadataStmt);
+    this.appendRemoteScopingStatements(ngModuleDef.additionalStatements, node, declarations);
+
+
```
One too many linebreaks? 👀

JoostK commented on packages/compiler/src/render3/partial/ng_module.ts:
```diff
@@ -0,0 +1,67 @@
+  definitionMap.set('type', meta.internalType);
+
+  // Only generate the keys in the metadata if the arrays have values.
+  // Note that for partial compilation we do not wrap forward-refs. This is handled in the linker.
+  // Hence passing `false` for `containsForwardDecls` in `refsToArray()` calls.
```
Oh, this probably explains my confusion around the `containsForwardDecls` flag. Why can the partial declaration get away with not wrapping forward decls into a function?

petebacondarwin replied:
I think... except perhaps in JIT mode.I'll add a test.

petebacondarwin replied:
Agh! OK. So you are right. It won't work in JIT mode. If it gets linked then it doesn't matter because the code gets transformed before it is executed.

petebacondarwin replied:
Hopefully I have solved this by unwrapping these functions when parsing the declaration so that we can re-wrap them in the definition!

JoostK commented on packages/compiler/src/render3/partial/ng_module.ts:
```diff
@@ -0,0 +1,67 @@
+  // Only generate the keys in the metadata if the arrays have values.
+  // Note that for partial compilation we do not wrap forward-refs. This is handled in the linker.
+  // Hence passing `false` for `containsForwardDecls` in `refsToArray()` calls.
+  if (meta.bootstrap.length) {
```
Can these be checked using `> 0`, please :)

JoostK commented on packages/compiler/src/render3/partial/ng_module.ts:
```diff
@@ -0,0 +1,67 @@
+    definitionMap.set('containsForwardDecls', o.literal(true));
+  }
+
+  if (meta.id) {
```
```suggestion
  if (meta.id !== undefined) {
```
or could be `!== null`

JoostK commented on packages/compiler/src/render3/partial/ng_module.ts:
```diff
@@ -0,0 +1,67 @@
+    definitionMap.set('exports', refsToArray(meta.exports, /* containsForwardDecls */ false));
+  }
+
+  if (meta.schemas && meta.schemas.length) {
```
```suggestion
  if (meta.schemas !== undefined && meta.schemas.length > 0) {
```

(or maybe it has to be `!== null`, I haven't looked up the actual type.)

petebacondarwin committed:
refactor(compiler): use `DefinitionMap` and remove `mapToMapExpression()`

By consistently using the `DefinitionMap` class there is no longer
any need for the `mapToMapExpression()` function.

petebacondarwin committed:
refactor(compiler): remove unused code

These functions are not used so can be removed.

petebacondarwin committed:
refactor(compiler): consolidate `R3CompiledExpression`

There were a number of almost identical interfaces used in
the same way throughout the Render3 compiler code.
This commit changes the compiler to use the same interface
throughout.

JoostK commented on packages/compiler/src/render3/r3_module_compiler.ts:
```diff
@@ -86,6 +86,43 @@ export interface R3NgModuleMetadata {
+   * An expression evaluating to an array of expressions representing the bootstrap components
+   * specified by the module.
+   */
+  bootstrap: o.Expression;
```
A few of these properies are optionally declared, so they should arguably be optional here too.

JoostK commented on packages/compiler/src/render3/partial/api.ts:
```diff
@@ -264,18 +258,66 @@ export interface R3DeclareQueryMetadata {
+   */
+  exports?: o.Expression[];
+
+  /**
+   * Whether to wrap the `bootstrap`, `declarations`, `imports` and `exports` into function calls to
+   * avoid referencing classes before they are declared.
+   */
+  containsForwardDecls?: boolean;
```
This property is no longer consumed during linker and I believe we should delete it from the generated code.

JoostK commented on packages/compiler/src/render3/partial/injector.ts:
```diff
@@ -0,0 +1,39 @@
+
+  definitionMap.set('type', meta.internalType);
+  definitionMap.set('providers', meta.providers);
+  if (meta.imports.length) {
```
```suggestion
if (meta.imports.length > 0) {
```

JoostK commented on packages/compiler/src/render3/util.ts:
```diff
@@ -25,6 +25,15 @@ export interface R3Reference {
+export interface R3CompiledExpression {
+  expression: o.Expression;
+  type: o.Type;
+  statements: o.Statement[];
```
Now that this is always here, it should arguably be used in the `DecoratorHandler.compile*` methods where a new empty array is currently passed. Otherwise it's quite strange that this field can be populated without it being used.

petebacondarwin committed:
refactor(compiler): consolidate `wrapReference()`

This function is declared in multiple places. The instances inside
`compiler` are slightly different to those in `compiler-cli`. So this
commit consolidates them into two reusable functions.

petebacondarwin committed:
refactor(compiler): move `type` in shared `R3PartialDeclaration` interface

All interfaces that extend `R3PartialDeclaration` require a `type`
property so it makes sense to put it in the base interface.

jessicajaniuk commented:
[Presubmit](https://test.corp.google.com/ui#id=OCL:362369810:BASE:362369858:1615499498340:82aacdb)

petebacondarwin commented:
Merge-assistance: the outstanding reviews are not required but Github is not allowing them to be removed.

petebacondarwin commented:
@JoostK pointed out that the JIT mode `ɵɵngDeclareNgModule()` implementation unwraps the forward ref wrapper functions too early. Working on a fix...

petebacondarwin committed:
fixup! refactor(compiler-cli): implement `ɵɵngDeclareNgModule` and `ɵɵngDeclareInjector`

petebacondarwin committed:
fixup! refactor(compiler-cli): implement `ɵɵngDeclareNgModule` and `ɵɵngDeclareInjector`

petebacondarwin committed:
docs(compiler): fix typos in partial compilation comments

petebacondarwin committed:
docs(core): fix typo in developer docs

JoostK commented on packages/compiler/src/render3/r3_module_compiler.ts:
```diff
@@ -181,6 +182,34 @@ export function compileNgModule(meta: R3NgModuleMetadata): R3CompiledExpression
+ * This function is used in JIT mode to generate the call to `ɵɵdefineNgModule()` from a call to
+ * `ɵɵngDeclareNgModule()`.
+ */
+export function compileNgModuleDeclarationExpression(meta: R3DeclareNgModuleFacade): o.Expression {
```
Although it's not ideal that the code generation now uses a completely separate codepath, I think this is a reasonable compromise (compared to changing `R3NgModuleMetadata` to have pre-compiled expressions instead of `R3Reference[]`.)

JoostK replied:
Funnily enough, this function achieves essentially this:

```ts
return o.importExpr(R3.defineNgModule).callFn([new o.WrappedNodeExpr(meta)]);
```

but I think the explicit form is desirable here; this is just an observation :-)

petebacondarwin replied:
> Funnily enough, this function achieves essentially this:
> 
> ```ts
> return o.importExpr(R3.defineNgModule).callFn([new o.WrappedNodeExpr(meta)]);
> ```
> 
> but I think the explicit form is desirable here; this is just an observation :-)

Indeed, as it turns out in this case! But I also feel that being explicit is cleaner here.

JoostK commented on packages/compiler/src/render3/r3_module_compiler.ts:
```diff
@@ -181,6 +182,34 @@ export function compileNgModule(meta: R3NgModuleMetadata): R3CompiledExpression
+  if (meta.id !== undefined) {
+    definitionMap.set('id', new o.WrappedNodeExpr(meta.id));
+  }
+  return o.importExpr(R3.defineNgModule).callFn([definitionMap.toLiteralMap()], undefined, true);
```
I'd remove the `undefined, true` args here as we don't care about the pure marker.

JoostK commented on packages/core/test/render3/jit/declare_ng_module_spec.ts:
```diff
@@ -7,6 +7,7 @@
  */
 
 import {NO_ERRORS_SCHEMA, ɵNgModuleDef, ɵɵngDeclareNgModule} from '@angular/core';
+import {SchemaMetadata, Type} from '@angular/core/src/core';
```
Move these imports into the `@angular/core` line above?

JoostK commented on packages/core/test/render3/jit/declare_ng_module_spec.ts:
```diff
@@ -101,10 +107,14 @@ function expectNgModuleDef(
 }
+
+function unwrap(values: Type<any>[]|(() => Type<any>[])): Type<any>[] {
+  return (typeof values === 'function') ? values() : values;
```
super-nit: unnecessary parens around the ternary condition

JoostK commented on packages/core/src/render3/definition.ts:
```diff
@@ -385,31 +385,8 @@ export const autoRegisterModuleById: {[id: string]: NgModuleType} = {};
-  /** Unique ID for the module that is used with `getModuleFactory`. */
-  id?: string | null;
-}): unknown {
+export function ɵɵdefineNgModule<T>(def: Pick<NgModuleDef<T>, 'type'>&
```
I think I prefer the prior situation, with the external, compiled type explicitly spelled out instead of some type juggling of the runtime type. Since the compiled shape and internal runtime shape are not necessarily the same thing, I'd keep their types separate as well.

petebacondarwin replied:
Yes you are right. I am not quite sure why I did this anyway 😺

JoostK reviewed (approved):
I think we've reached the finishing line now 🏁

JoostK commented on packages/compiler-cli/src/ngtsc/annotations/src/ng_module.ts:
```diff
@@ -6,7 +6,8 @@
 
-import {compileInjector, compileNgModule, CUSTOM_ELEMENTS_SCHEMA, Expression, ExternalExpr, Identifiers as R3, InvokeFunctionExpr, LiteralArrayExpr, LiteralExpr, NO_ERRORS_SCHEMA, R3DependencyMetadata, R3FactoryTarget, R3Identifiers, R3InjectorMetadata, R3NgModuleMetadata, R3Reference, SchemaMetadata, Statement, STRING_TYPE, WrappedNodeExpr} from '@angular/compiler';
+import {compileDeclareInjectorFromMetadata, compileDeclareNgModuleFromMetadata, compileFactoryFunction, compileInjector, compileNgModule, CUSTOM_ELEMENTS_SCHEMA, Expression, ExternalExpr, Identifiers as R3, InvokeFunctionExpr, LiteralArrayExpr, LiteralExpr, NO_ERRORS_SCHEMA, R3FactoryMetadata, R3FactoryTarget, R3Identifiers, R3InjectorMetadata, R3NgModuleMetadata, R3Reference, SchemaMetadata, Sta ...
+import {R3CompiledExpression} from '@angular/compiler/src/compiler';
```
Can this go into `@angular/compiler` as well (may require a public export)

AndrewKushnir commented:
[Presubmit](http://test/OCL:362869535:BASE:362869551:1615787896089:bd700890).

AndrewKushnir commented:
FYI, presubmits are successful for the changes in this PR.

Note to Caretaker: GitHub UI and NgBot indicate 3 pending code reviews, but it looks like everyone on that list already reviewed and approved these changes, so this PR should be ready to go.

jessicajaniuk commented:
There are conflicts with `11.2.x` branch. So this has been rebased and CI is re-running. I'll try to merge again after CI finishes, but we might need a patch-only PR.

jessicajaniuk commented:
@petebacondarwin The merge conflicts with `11.2.x` still exist after rebase. I'm switching this to `target: minor` and merging. Please create a patch-only PR to get this fully merged into `11.2.x`.

petebacondarwin commented:
Thanks @jessicajaniuk - since the next PR in the sequence is major-only anyway, and all PRs following that one will only go in major, there is not much benefit in this one landing in patch by itself. So I think we are good.
